rook-operator: panic: unexpected long blocking (> 1 Minute) when handling cluster event

I noticed the rook-operator pod had 1 restart:
```
> kubectl get pod
NAME                              READY     STATUS    RESTARTS   AGE
rook-operator-2761695524-th25c    1/1       Running   1          10m
wordpress-255188236-ns24k         1/1       Running   0          2m
wordpress-mysql-830254888-jldzd   1/1       Running   0          2m
```

Looking at the previous pod instance logs, we see a panic:
```
> kubectl logs --previous=true rook-operator-2761695524-th25c
2017-02-27 23:52:58.931447 I | rook-operator: starting operator in namespace default
2017-02-27 23:52:58.931645 I | operator: creating rook TPR
2017-02-27 23:53:07.969332 I | operator: start watching rook tpr: 0
2017-02-27 23:53:07.971781 I | operator: start watching at 0
2017-02-27 23:53:07.973595 I | operator: starting cluster. &{Namespace:rook factory:0xc4f74e0 clientset:0xc4200489a0 Metadata:{Name: GenerateName: Namespace: SelfLink: UID: ResourceVersion: Generation:0 CreationTimestamp:0001-01-01 00:00:00 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Spec:{Namespace: Version:dev-2017-02-27-k8s Paused:false UseAllDevices:false}}
2017-02-27 23:53:07.981285 I | op-mon: start running mons
2017-02-27 23:53:07.984505 I | op-mon: creating mon secrets for a new cluster
2017-02-27 23:53:08.025627 I | op-mon: saved rook-admin secret
2017-02-27 23:53:08.030518 I | op-mon: there are 2 nodes available for 3 monitors
2017-02-27 23:53:08.035853 I | op-mon: 0 running, 0 pending pods
2017-02-27 23:53:08.050634 I | op-mon: waiting 6s for pod mon0 to start. status=
2017-02-27 23:53:14.056081 I | op-mon: waiting 6s for pod mon0 to start. status=
2017-02-27 23:53:20.061363 I | op-mon: waiting 6s for pod mon0 to start. status=
2017-02-27 23:53:26.066015 I | op-mon: waiting 6s for pod mon0 to start. status=
2017-02-27 23:53:32.073562 I | op-mon: pod mon0 started
2017-02-27 23:53:32.090926 I | op-mon: waiting 6s for pod mon1 to start. status=
2017-02-27 23:53:38.097995 I | op-mon: waiting 6s for pod mon1 to start. status=
2017-02-27 23:53:44.105030 I | op-mon: waiting 6s for pod mon1 to start. status=
2017-02-27 23:53:50.110483 I | op-mon: waiting 6s for pod mon1 to start. status=
2017-02-27 23:53:56.115595 I | op-mon: pod mon1 started
2017-02-27 23:53:56.130365 I | op-mon: waiting 6s for pod mon2 to start. status=
2017-02-27 23:54:02.138551 I | op-mon: pod mon2 started
2017-02-27 23:54:02.138649 I | op-mon: started 3/3 mons (0 already running)
2017-02-27 23:54:02.138698 I | op-api: starting the Rook api
2017-02-27 23:54:02.154622 I | op-api: API service running at 10.3.0.115:8124
2017-02-27 23:54:02.177650 I | op-api: api deployment started
2017-02-27 23:54:02.177704 I | op-osd: start running osds
2017-02-27 23:54:02.225130 I | op-osd: osd daemon set started
2017-02-27 23:54:02.225346 I | cephmon: connecting to ceph cluster rookcluster with user admin
panic: unexpected long blocking (> 1 Minute) when handling cluster event

goroutine 97 [running]:
panic(0x2da8dc0, 0xc4201b5e80)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/rook/rook/pkg/operator/k8sutil.(*PanicTimer).Start.func1()
	/home/rook/go/src/github.com/rook/rook/pkg/operator/k8sutil/timer.go:39 +0x5d
created by time.goFunc
	/usr/local/go/src/time/sleep.go:154 +0x44
```

It seems like if the cluster creation takes longer than 1 minute, we would hit this.  Perhaps the 1 minute panic is too aggressive for the handling of a cluster "Added" event?

--- a/pkg/operator/operator.go
+++ b/pkg/operator/operator.go
@@ -102,17 +102,15 @@ func (o *Operator) initResources() (string, error) {
 
 	err = o.createTPR()
 	if err != nil {
-		if k8sutil.IsKubernetesResourceAlreadyExistError(err) {
-			// TPR has been initialized before. We need to recover existing cluster.
-			watchVersion, err := o.findAllClusters()
-			if err != nil {
-				return "", fmt.Errorf("failed to find clusters. %+v", err)
-			}
-			return watchVersion, nil
-		}
+		return "", fmt.Errorf("failed to create TPR. %+v", err)
 	}
 
-	return "0", nil
+	// Check if there is an existing cluster to recover
+	watchVersion, err := o.findAllClusters()
+	if err != nil {
+		return "", fmt.Errorf("failed to find clusters. %+v", err)
+	}
+	return watchVersion, nil
 }
 
 func (o *Operator) watchTPR(watchVersion string) error {
@@ -147,15 +145,12 @@ func (o *Operator) watchTPR(watchVersion string) error {
 				} else {
 					newCluster := newCluster(ns, c.Spec.Version, c.Spec.UseAllDevices, o.factory, o.clientset)
 					stopCh := make(chan struct{})
-					logger.Infof("starting cluster. %+v", newCluster)
-
 					o.stopChMap[ns] = stopCh
 					o.clusters[ns] = newCluster
 					o.clusterRVs[ns] = c.Metadata.ResourceVersion
-					err := newCluster.CreateInstance()
-					if err != nil {
-						logger.Errorf("failed to create cluster %s. %+v", ns, err)
-					}
+
+					logger.Infof("starting new cluster %s in namespace %s", c.Metadata.Name, ns)
+					go startCluster(newCluster)
 				}
 
 			case kwatch.Modified:
@@ -172,22 +167,32 @@ func (o *Operator) watchTPR(watchVersion string) error {
 
 }
 
+func startCluster(cluster *Cluster) {
+	err := cluster.CreateInstance()
+	if err != nil {
+		logger.Errorf("failed to create cluster in namespace %s. %+v", cluster.Namespace, err)
+	}
+}
+
 func (o *Operator) findAllClusters() (string, error) {
 	logger.Info("finding existing clusters...")
 	clusterList, err := getClusterList(o.clientset.CoreV1().RESTClient(), o.Namespace)
 	if err != nil {
 		return "", err
 	}
-
+	logger.Infof("found %d clusters", len(clusterList.Items))
 	for i := range clusterList.Items {
 		c := clusterList.Items[i]
 
 		stopCh := make(chan struct{})
 		ns := c.Spec.Namespace
-		newCluster := newCluster(ns, c.Spec.Version, c.Spec.UseAllDevices, o.factory, o.clientset)
+		existingCluster := newCluster(ns, c.Spec.Version, c.Spec.UseAllDevices, o.factory, o.clientset)
 		o.stopChMap[ns] = stopCh
-		o.clusters[ns] = newCluster
+		o.clusters[ns] = existingCluster
 		o.clusterRVs[ns] = c.Metadata.ResourceVersion
+
+		logger.Infof("resuming cluster %s in namespace %s", c.Metadata.Name, ns)
+		go startCluster(existingCluster)
 	}
 
 	return clusterList.Metadata.ResourceVersion, nil
@@ -216,8 +221,6 @@ func (o *Operator) watch(watchVersion string) (<-chan *Event, <-chan error) {
 				errCh <- errors.New("invalid status code: " + resp.Status)
 				return
 			}
-
-			logger.Infof("start watching at %v", watchVersion)
 
 			decoder := json.NewDecoder(resp.Body)
 			for {

--- a/pkg/operator/tpr.go
+++ b/pkg/operator/tpr.go
@@ -54,7 +54,9 @@ func (o *Operator) createTPR() error {
 	}
 	_, err := o.clientset.ExtensionsV1beta1().ThirdPartyResources().Create(tpr)
 	if err != nil {
-		return fmt.Errorf("failed to create rook third party resources. %+v", err)
+		if !k8sutil.IsKubernetesResourceAlreadyExistError(err) {
+			return fmt.Errorf("failed to create rook third party resources. %+v", err)
+		}
 	}
 
 	return o.waitForTPRInit(o.clientset.CoreV1().RESTClient(), 3*time.Second, 90*time.Second, o.Namespace)

